### PR TITLE
feat(ui): Project API integration

### PR DESCRIPTION
**Pull request by jashanbhullar**
# Overview
Project API integration

## What I've done
- graphql queries 
- hook for executing queries 
- Get/Create/Delete project hooks 

## What I haven't done
- Update Project - Check `Notes` section 

## How I tested
- Manually

## Which point I want you to review particularly
- graphql queries
- hook implementation 
- react-query implementation

### Notes:
- Dialog component should be able to return a `boolean` value for `confirm` dialogs (Need this for `delete` projects/workspaces). Should revisit Dialog implementation before adding more components on it 
- The dialog component doesn't stop the navigation events, user can open a dialog and still change pages with the keyboard (should stop all events not related to the dialog)  

### Improvements:
- No pagination implementation as of now 
- Not sure of the proper return structure of the data (data, meta, ...rest [isLoading, isError...so on]), is this okay? Is the react-query abstracted enough?

### CODE REVIEW

File: ui/src/features/Dashboard/components/MainSection.tsx [hunk 1]
```diff
@@ -15,16 +16,21 @@ import {
...
 import type { Project } from "@flow/types";
+import { formatDate } from "@flow/utils";
 
 const MainSection: React.FC = () => {
   const t = useT();
   const [currentWorkspace] = useCurrentWorkspace();
   const [currentProject, setCurrentProject] = useCurrentProject();
   const navigate = useNavigate({ from: "/workspace/$workspaceId" });
+  const { useGetProjects, deleteProject } = useProject();
+  const { projects } = useGetProjects(currentWorkspace?.id as string);
```
Comment (KaWaite) [inline ui/src/features/Dashboard/components/MainSection.tsx:32]: As discussed on Slack, do not typecast a type on a possible undefined.

File: ui/src/features/Dashboard/components/MainSection.tsx [hunk 2]
```diff
@@ -33,7 +39,17 @@ const MainSection: React.FC = () => {
...
+  useEffect(() => {
+    if (!projects) return;
+    projects.forEach(p => {
+      p.workflow = generateWorkflows(1)[0];
+    });
+  }, [projects]);
+
+  const handleDeleteProject = async (id: string) => {
+    await deleteProject(id);
+  };
```
Comment (KaWaite) [inline ui/src/features/Dashboard/components/MainSection.tsx:52]: Keep in mind we will eventually want a warning Dialog for actions like this. (once you read this you can resolve this comment yourself 👍🏻 )

File: ui/src/lib/gql/project/hook.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    useQuery({
+      queryKey: [ProjectQueryKeys.GetProjects, workspaceId],
+      queryFn: async () => {
+        // TODO: Should have pagination
+        const data = await graphQLContext?.GetProjects({ workspaceId, first: 20 });
+        if (!data) return {};
+        const {
+          projects: { nodes, ...rest },
+        } = data;
+        return { projects: nodes as Project[], meta: rest };
```
Comment (KaWaite) [inline ui/src/lib/gql/project/hook.ts:38]: Don't typecast here either. projects will still have all of the variables that come from the query, but will just be invisible to everyone, which isn't ideal.
Instead, map through nodes and create new objects that match `Project`.
That way everything is consistent and we can be more deliberate what we are sending to the app 🙏🏻

File: ui/src/features/Dashboard/components/MainSection.tsx (hunk 2 above)
Comment (jashanbhullar) [reply ui/src/features/Dashboard/components/MainSection.tsx:52]: Added a todo for reference. Same as workspace

File: ui/src/features/Dashboard/components/MainSection.tsx (hunk 1 above)
Comment (jashanbhullar) [reply ui/src/features/Dashboard/components/MainSection.tsx]: I passed it as a prop since it needs to exist before we can render this component.

Comment (jashanbhullar) [reply ui/src/features/Dashboard/components/MainSection.tsx]: We are setting the state in the parent. I am not sure how you feel about using global state at too many places?

File: ui/src/lib/gql/project/hook.ts (hunk 3 above)
Comment (jashanbhullar) [reply ui/src/lib/gql/project/hook.ts]: This is Done!

File: ui/src/features/Dashboard/components/MainSection.tsx (hunk 1 above)
Comment (KaWaite) [reply ui/src/features/Dashboard/components/MainSection.tsx]: SETTING global state should be only in very deliberate places so its easy to track. 
USING global state I think can, and should, be used liberally. That way we can avoid prop drilling and components are more mobile if we chose to rearrange the UI, etc

**Review by KaWaite — APPROVED** [review]

File: ui/src/lib/gql/project/hook.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+export const useProject = () => {
+  const graphQLContext = useGraphQLContext();
+  const queryClient = useQueryClient();
+
+  const createProjectMutation = useMutation({
+    mutationFn: async (input: CreateProjectInput) => {
+      const data = await graphQLContext?.CreateProject({ input });
+      return { project: data?.createProject?.project };
```
Comment (KaWaite) [inline ui/src/lib/gql/project/hook.ts:19]: Sorry realized you are passing the gql project here

File: ui/src/lib/gql/project/hook.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      queryKey: [ProjectQueryKeys.GetProjects, workspaceId],
+      queryFn: async () => {
+        // TODO: Should have pagination
+        const data = await graphQLContext?.GetProjects({ workspaceId, first: 20 });
+        if (!data) return {};
+        const {
+          projects: { nodes, ...rest },
+        } = data;
+
+        const projects: Project[] = nodes.flatMap(f => (f ? [f] : []));
```
Comment (KaWaite) [inline ui/src/lib/gql/project/hook.ts:39]: Sorry realized you are passing the gql project here too

File: ui/src/lib/gql/project/hook.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      queryClient.invalidateQueries({
+        queryKey: [ProjectQueryKeys.GetProjects, workspaceId],
+      }),
+  });
+
+  const createProject = async (input: CreateProjectInput): Promise<CreateProject> => {
+    const { mutateAsync, ...rest } = createProjectMutation;
+    try {
+      const data = await mutateAsync(input);
+      return { project: data.project, ...rest };
```
Comment (KaWaite) [inline ui/src/lib/gql/project/hook.ts:62]: And here

File: ui/src/lib/gql/project/hook.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    } catch (err) {
+      return { project: undefined, ...rest };
+    }
+  };
+
+  const useGetProjects = (workspaceId: string): GetProjects => {
+    const { data, ...rest } = useGetProjectsQuery(workspaceId);
+    return {
+      projects: data?.projects,
```
Comment (KaWaite) [inline ui/src/lib/gql/project/hook.ts:71]: And here

**Review by KaWaite — CHANGES_REQUESTED** [review]
Sorry noticed something

File: ui/src/lib/gql/project/hook.ts (hunk 4 above)
Comment (jashanbhullar) [reply ui/src/lib/gql/project/hook.ts:19]: image
I am using the one from types!

**Review by KaWaite — APPROVED** [review]

**Pull request merged**